### DISCUSSION TITLE
Progress Reports

### DISCUSSION THREAD

**Discussion by statmike** [statmike/vertex-ai-mlops#9] [Announcements]
## Starting this discussion as a way to communicate where the content is growing and refining currently and in the near future

**Comment by statmike** [comment]
## July & August 2022
The shape of this repository changed a lot over the past 2 months (July, August) in preparation for expansion.  The numbered folders will all approach the same supervised classification model with different frameworks (autoML, BQML, sklearn, TensorFlow, R, XGBoost, ...).  The content within those folders will adjust to workflows for that framework on Vertex AI while incorporating the features of Vertex AI in layers.  The [05 - TensorFlow](../tree/main/05%20-%20TensorFlow) folder is the first to get the full revamp and is a great place to follow progress.

**Comment by statmike** [comment]
## September 2022
September started with a full revamp of the [03 - BigQuery ML (BQML)](../tree/main/03%20-%20BigQuery%20ML%20(BQML)) series to include examples of all supervised classification methods with full model journey examples.  The content was expanded with examples for hyperparameter tuning and a new pipeline showing a process to continually update an endpoint with the best overall model.

The past week has been devoted to a new series called [Tips](../tree/main/Tips).  These are the baseline for many of the techniques that will be used in the content going forward.  There is now a full series of tips covering
- scaling jobs with parallelism
- parameter input options for command line (argparse, docopt, click) and files (JSON, YAML, pickle)
- Use the Python Client for GCS to interact with files
- Packaging Python code for portability: files/modules, folders of modules, source distributions, built distributions, Artifact Registry
- Building Custom Containers that incorporate code with the software of choice using Cloud Build 
- Workflows for pushing compute out of the notebook instance and into Vertex AI Training Custom Jobs with a focus on still testing jobs locally in a notebooks.  This brings together a developer workflow for ML.

**Comment by statmike** [comment]
## Q4 2022 Plans:
A view towards the near future.  These are my plans for Q4 and early 2023. 

### September & October 2022
**What is happening next?**

I will be bouncing between:
- A refinement pass for the 05 series before expanding it with ML Metadata and more examples
- A small refinement pass for the 03 series and completing pipeline 2 and 3
- A full refinement pass for the 02 series as a baseline to expand this for AutoML Tabular Workflows
- A big refinement pass for the Applied Forecasting series that makes it more flexible for adapting to users data with many more series than the example has.  Also, adding a new graphing technique that makes reviewing time series more interactive.

### Horizon 1: Late October -> 2022
**What is on the horizon?**

As the content is 02, 03, 05, and Applied Forecasting catch up to the new standard I will turn my focus to these areas:
- Back to filming tutorial walkthroughs of the content
- Continued expansion of the content based on feedback
- Incorporating a full feature store example
- Expand 02 with AutoML Tabular Workflows
- Building out 04, 06, 07, 08 in the form of the new 05 layout
     - Getting help from friends!
- Adding options for scripting GCP project creation for this work: from gcloud to terraform
     - Getting help from friends!

### Horizon 2: 2023 start
- migrating more projects into this repository:
     - A text example series with BERT that also shows dataquality with BigQuery Remote Functions and serverless Dataproc
     - A Matching Engine project
     - Pre-Built API examples (moving these from Gist into the repository)
     - Solutions examples like Document AI (moving these from Gist into the repository)
     - Autoencoders! a favorite topic

**Comment by statmike** [comment]
## October 2022
A big month of updates! 

Highlights:
- Multiple refinement passes for the `05 - TensorFlow` series.   
- A demo of automating a Vertex AI jobs (training, pipelines, predictions, ...).  The `05Tools - Automation` demo uses a Pub/Sub message to trigger a Cloud Function that start a Vertex AI job - a training job in the demo but could be any job!  The architecture inclusion of PubSub means the process can be triggered from anywhere: manual, scheduled, event based.  In the notebook both manual and scheduled (Cloud Scheduler) are used.
- Many new workflows for the R language in `08 - R`
    - A notebook based example using an R kernel
    - A example of building an R script for running in Vertex AI Training Jobs
    - A complete workflow from script to Vertex AI Training with a custom container, then serving in both Vertex AI with online Endpoints
    - A workflow for serving in Cloud Run and locally on the users machine. 
    - An example of moving large BigQuery table to R dataframe by first exporting to multiple files in GCS and then using the GCS FUSE mount in Vertex AI Training jobs to read into R
    - A complete workflow from script creation to custom container to Vertex AI Training and Serving
- A full refinement pass for `03 - BigQuery ML (BQML)`
    - Fixed Pipeline Example 2 to follow the summers updates to the repository
    - Added `03g` example - PCA including using anomalies to classify potential fraud
    - Added `03h` example - k-means including using anomalies to classify potential fraud
    - Added `03i` example - Autoencoder with Hyperparameter Tuning for architecture and using anomalies to classify potential fraud
- Started a refinement pass for `Applied Forecasting`
    - Notebooks 1 and 2 now make it easier to change the data source to your data.  All the visuals limit the number of time series displayed based on a parameter set at the top `viz_limit`
-   New examples in `Applied Pre-Trained APIs` included a complete workflow with Cloud Vision, Cloud Document AI, Cloud Natural Language, Cloud Test to Speech, Cloud Speech to Text, and Cloud Translation.
-  Revision to the main readme for clarity and many links for Vertex AI pre-trained APIs and AutoML offerings
- Many revisions and additions to the `Tips` folder
- The start of a scripting project that gives user options for automating the setup of this repository for learning and demonstration

Planned items pushed to November:
 - Continue refinement/update pass for the 02 Vertex AI AutoML - with new additions
 - Continue refinement/update pass for the Applied Forecasting - with new additions
 - Start a new project for Julia
 - Push the first sklearn examples
 - MORE
